README.md
# ThinkVue
--------------
## 安装方法



## 注意的地方：

1. 使用请求头authKey和sessionId验证权限，关闭了cookie；
2. 在全局配置文件config/app.php中
3. 使用严格路由，未声明的地址重定位至 `admin/base/miss` ，可在`route/route.php`中取消。

## 投入生产环境应做如下更改：

1. 更改日志文件级别（在全局配置文件/config/app.php中）
2. 关闭调试模式（/config/app.php=>app_debug）
3. 更改数据库密码（/config/database.php）
4. 如果在开发环境中关闭了缺省路由，则应注意开启缺省路由

## 返回格式和约定返回值

返回格式为 ` json([code=>int,data=>object,error=>msg]) ` ，其中code为返回值，大于0为正确返回，小于0为错误返回。1为默认返回，-1为默认错误
* -1001~-1999 用户权限信息错误
    + -1001 登录失效
    + -1002 没有操作权限
    + -1003 账号或者密码错误
    + -1004 修改的新密码与原密码一致（准备废弃，用-2002代替）
    + -1005 该客户端未绑定账号
    + -1006 账号被禁用
* -2001~-2999 系统错误
    + -2001 发生了严重错误导致无法进行
    + -2002 参数错误
* -3001~-3999 数据错误
    + -3001 数据结构错误
    + -3002 数据存在冲突
    + -3003 数据库未知错误
    + -3004 指定数据不存在
* -4001~-4999 配置错误
    + -4001 地址错误
    + -4002 第三方OAuth2.0登录错误
    + -4003 服务器网络错误

CHANGELOG.md
## V5.1.39 LTS（2019-11-18）

本次更新为常规更新，主要包括：

* 修正`memcached`驱动
* 改进`HasManyThrough`关联查询
* 改进`Request`类`isJson`方法
* 改进关联查询
* 改进`redis`驱动
* 增加 Model类`getWhere`方法对复合主键的支持
* 改进`newQuery`方法
* 改进闭包查询的参数绑定
* 修正`Validate`
* 修复某些情况下URL会多一个冒号
* 调整composer.json
* 修复使用`Cache::clear()`时，报错缓存文件不存在问题
* 使用File类的unlink方法进行文件删除
* 改进`paraseData`方法
* 修正image验证方法
* 改进Url生成
* 改进空操作对数字的支持
* 改进一处PHP7.4兼容性问题

## V5.1.38 LTS（2019-8-8）

本次更新为常规更新，主要包括：

* `Request`类增加`isJson`方法
* 改进浮点型查询
* 修正关联查询关联外键为空的查询错误
* 远程一对多支持关联统计和预载入查询
* 远程一对多关联支持`has`/`hasWhere`查询
* 优化`parseIn`解析
* 改进`parseLike`查询
* 改进Url生成
* 改进模型的`toArray`方法
* 修正`notIn`查询
* 改进`JSON`字段查询
* 改进Controller类`display`/`fetch`方法返回`ViewResponse`对象
* 改进`param`方法
* 改进`mysql`驱动`getExplain`方法
* 改进时间查询
* 改进模型关联的`has`/`hasWhere`方法对软删除的支持
* 修正社区反馈的BUG

## V5.1.37 LTS（2019-5-26）

本次更新为常规更新，主要更新如下：

* 改进关联数据更新
* 修正关联动态获取器
* 改进`redis`驱动 
* 修复验证规则里面出现二维数组时的错误
* 改进跨域请求支持
* 完善模型`hidden`方法对关联属性的支持
* 改进`where`查询方法传入`Query`对象的支持`bind`数据
* 改进数据集对象的`load`方法
* 修正缓存类`clear`方法对`tag`的支持

## V5.1.36 LTS（2019-4-28）

本次更新为常规更新，主要更新如下：

* 修正`chunk`方法一处异常抛出的错误
* 修正模型输出的`visible`
* 改进环境变量加载
* 改进命令行日志的`level`配置支持
* 修复设置有缓存前缀时，无法清空缓存标签的问题
* HasMony对象`saveAll`方法兼容`Collection`格式参数格式
* 修正`whereOr`查询使用字符串的问题
* 改进`dateFormat`设置对写入数据的影响
* 修正查询缓存
* 记住指定的跳转地址
* 改进软删除
* 改进聚合查询SQL去除limit 1
* 改进缓存驱动

## V5.1.35 LTS（2019-3-2）

本次主要为常规更新，修正了一些反馈的问题。

* 修正验证类自定义验证方法执行两次的问题
* 模型增加`isEmpty`方法用于判断是否空模型
* 改进获取器对`append`的支持
* 修正一对多关联的`withCount`自关联问题
* facade类注释调整
* 改进关联属性的`visible`和`hidden`判断
* 修正路由分组的`MISS`路由
* 改进pgsql.sql

## V5.1.34 LTS（2019-1-30）

本次更新为常规更新，修正了一些反馈的问题。

* 改进Request类的`has`方法，支持`patch`
* 改进`unique`验证的多条件支持
* 修复自定义上传验证，检测文件大小
* 改进`in`查询支持表达式
* 改进路由的`getBind`方法
* 改进验证类的错误信息获取
* 改进`response`助手函数默认值
* 修正mysql的`regexp`查询
* 改进模型类型强制转换写入对`Expression`对象的支持

## V5.1.33 LTS（2019-1-16）

* 修复路由中存在多个相同替换的正则BUG
* 修正whereLike查询
* join方法支持参数绑定
* 改进union方法
* 修正多对多关联的attach方法
* 改进验证类的正则规则自定义
* 改进Request类method方法
* 改进File日志类型的CLI日志写入
* 改进文件日志time_format配置对JSON格式的支持

## V5.1.32 LTS（2018-12-24）

本次主要为常规更新，修正了一些反馈的问题。


* 改进多对多关联的`attach`方法 
* 改进聚合查询的`field`处理
* 改进关联的`save`方法
* 修正模型`exists`方法返回值
* 改进时间字段写入和输出
* 改进控制器中间件的调用
* 改进路由变量替换的性能
* 改进缓存标签的处理机制

## V5.1.31 LTS (2018-12-9)

本次版本包含一个安全更新，建议升级。

* 改进`field`方法
* 改进`count`方法返回类型
* `download`函数增加在浏览器中显示文件功能
* 修正多对多模型的中间表数据写入
* 改进`sqlsrv`驱动支持多个Schemas模式查询
* 统一助手函数与\think\facade\Response\Download函数文件过期时间
* 完善关联模型的`save`方法 增加`make`方法仅创建对象不保存
* 修改条件表达式对静态变量的支持
* 修正控制器名获取
* 改进view方法的`field`解析

## V5.1.30 LTS（2018-11-30）

该版本为常规更新，修正了一些社区反馈的问题。

主要更新如下：

* 改进查询类的`execute`方法
* 判断路由规则定义添加对请求类型的判断
* 修复`orderRaw`异常
* 修正 `optimize:autoload`指令
* 改进软删除的`destroy`方法造成重复执行事件的问题
* 改进验证类对扩展验证规则 始终验证 不管是否`require`
* 修复自定义验证`remove`所有规则的异常
* 改进时间字段的自动写入支持微秒数据
* 改进`Connection`类的`getrealsql`方法
* 修正`https`地址的URL生成
* 修复 `array_walk_recursive` 在低于PHP7.1消耗内部指针问题
* 改进手动参数绑定使用
* 改进聚合查询方法的`field`参数支持`Expression`

## V5.1.29 LTS（2018-11-11）

该版本主要改进了参数绑定的解析问题和提升性能，并修正了一些反馈的问题。

* 改进手动参数绑定
* 修正MISS路由的分组参数无效问题
* 行为支持对象的方法
* 修正全局查询范围
* 改进`belongsto`关联的`has`方法
* 改进`hasMany`关联
* 改进模型观察者多次注册的问题
* 改进`query`类的默认查询参数处理
* 修正`parseBetween`解析方法
* 改进路由地址生成的本地域名支持
* 改进参数绑定的实际URL解析性能
* 改进`Env`类的`getEnv`和`get`方法
* 改进模板缓存的生成优化
* 修复验证类的多语言支持
* 修复自定义场景验证`remove`规则异常
* File类添加是否自动补全扩展名的选项
* 改进`strpos`对子串是否存在的判断
* 修复`choice`无法用值选择第一个选项问题
* 验证器支持多维数组取值验证
* 改进解析`extend`和`block`标签的正则

## V5.1.28 LTS（2018-10-29）

该版本主要修正了上一个版本存在的一些问题，并改进了关联查询

* 改进聚合查询方法的字段支持DISTINCT
* 改进定义路由后url函数的端口生成
* 改进控制器中间件对`swoole`等的支持
* 改进Log类`save`方法
* 改进验证类的闭包验证参数
* 多对多关联支持指定中间表数据的名称
* 关联聚合查询支持闭包方式指定聚合字段
* 改进Lang类`get`方法
* 多对多关联增加判断关联数据是否存在的方法
* 改进关联查询使用`fetchsql`的情况
* 改进修改器的是否已经执行判断
* 增加`afterWith`和`beforeWith`验证规则 用于比较日期字段

## V5.1.27 LTS（2018-10-22）

该版本主要修正了路由绑定的参数，改进了修改器的执行多次问题，并正式宣布为LTS版本！


* 修正路由绑定的参数丢失问题
* 修正路由别名的参数获取
* 改进修改器会执行多次的问题

## V5.1.26（2018-10-12）

该版本主要修正了上一个版本的一些问题，并改进了全局查询范围的支持，同时包含了一个安全更新。


* 修正单一模块下注解路由无效的问题
* 改进数据库的聚合查询的字段处理
* 模型类增加`globalScope`属性定义 用于指定全局的查询范围
* 模型的`useGlobalScope`方法支持传入数组 用于指定当前查询需要使用的全局查询范围
* 改进数据集的`order`方法对数字类型的支持
* 修正上一个版本`order`方法解析的一处BUG
* 排序字段不合法或者错误的时候抛出异常
* 改进`Request`类的`file`方法对上传文件的错误判断

##  V5.1.25（2018-9-21）

该版本主要改进了查询参数绑定的性能和对浮点型的支持，以及一些细节的完善。

* 修正一处命令行问题
* 改进`Socketlog`日志驱动，支持自定义默认展开日志类别
* 修正`MorphMany`一处bug
* 跳转到上次记住的url，并支持默认值
* 改进模型的异常提示
* 改进参数绑定对浮点型的支持
* 改进`order`方法解析
* 改进`json`字段数据的自动编码
* 改进日志`log_write`可能造成的日志写入死循环
* Log类增加`log_level`行为标签位置，用于对某个类型的日志进行处理
* Route类增加`clear`方法清空路由规则
* 分布式数据库配置支持使用数组
* 单日志文件也支持`max_files`参数
* 改进查询参数绑定的性能
* 改进别名路由的URL后缀参数检测
* 控制器前置方法和控制器中间件的`only`和`except`定义不区分大小写

## V5.1.24（2018-9-5）

该版本主要增加了命令行的表格输出功能，并增加了查看路由定义的指令，以及修正了社区的一些反馈问题。

* 修正`Request`类的`file`方法
* 修正路由的`cache`方法
* 修正路由缓存的一处问题
* 改进上传文件获取的异常处理
* 改进`fetchCollection`方法支持传入数据集类名
* 修正多级控制器的注解路由生成
* 改进`Middleware`类`clear`方法
* 增加`route:list`指令用于[查看定义的路由](752690) 并支持排序
* 命令行增加`Table`输出类
* `Command`类增加`table`方法用于输出表格
* 改进搜索器查询方法支持别名定义
* 命令行配置增加`auto_path`参数用于定义自动载入的命令类路径
* 增加`make:command`指令用于[快速生成指令](354146)
* 改进`make:controller`指令对操作方法后缀的支持
* 改进命令行的定义文件支持索引数组 用于指令对象的惰性加载
* 改进`value`和`column`方法对后续查询结果的影响
* 改进`RuleName`类的`setRule`方法

## V5.1.23（2018-8-23）

该版本主要改进了数据集对象的处理，增加了`findOrEmpty`方法，并且修正了一些社区反馈的BUG。

* 数据集类增加`diff`/`intersect`方法用于获取差集和交集（默认根据主键值比较）
* 数据集类增加`order`方法支持指定字段排序
* 数据集类增加`map`方法使用回调函数处理数据并返回新的数据集对象
* Db增加`allowEmpty`方法允许`find`方法在没有数据的时候返回空数组或者空模型对象而不是null
* Db增加`findOrEmpty`方法
* Db增加`fetchCollection`方法用于指定查询返回数据集对象
* 改进`order`方法的数组方式解析，增强安全性
* 改进`withSearch`方法，支持第三个参数传入字段前缀标识，用于多表查询字段搜索
* 修正`optimize:route`指令开启类库后缀后的注解路由生成
* 修正redis缓存及session驱动
* 支持指定`Yaconf`的独立配置文件
* 增加`yaconf`助手函数用于配置文件


## V5.1.22（2018-8-9）

该版本主要增加了模型搜索器和`withJoin`方法，完善了模型输出和对`Yaconf`的支持，修正了一些社区反馈的BUG。

* 改进一对一关联的`table`识别问题
* 改进内置`Facade`类
* 增加`withJoin`方法支持`join`方式的[一对一关联](一对一关联.md)查询
* 改进`join`预载入查询的空数据问题
* 改进`Config`类的`load`方法支持快速加载配置文件
* 改进`execute`方法和事务的断线重连
* 改进`memcache`驱动的`has`方法
* 模型类支持定义[搜索器](搜索器.md)方法
* 完善`Config`类对`Yaconf`的支持
* 改进模型的`hidden/visible/append/withAttr`方法，支持在[查询前后调用](数组访问.md)，以及支持数据集对象
* 数据集对象增加`where`方法根据字段或者关联数据[过滤数据](模型数据集.md)
* 改进AJAX请求的`204`判断


## V5.1.21（2018-8-2）


application/common.php
<?php
// +----------------------------------------------------------------------
// | ThinkPHP [ WE CAN DO IT JUST THINK ]
// +----------------------------------------------------------------------
// | Copyright (c) 2006-2016 http://thinkphp.cn All rights reserved.
// +----------------------------------------------------------------------
// | Licensed ( http://www.apache.org/licenses/LICENSE-2.0 )
// +----------------------------------------------------------------------
// | Author: 流年 <liu21st@gmail.com>
// +----------------------------------------------------------------------

// 应用公共文件

use AlibabaCloud\Client\AlibabaCloud as AlibabaCloud;
use AlibabaCloud\Client\Exception\ClientException;
use AlibabaCloud\Client\Exception\ServerException;


/**
 * 行为绑定
 */
\think\facade\Hook::add('app_init', 'app\\common\\behavior\\InitConfigBehavior');

// +-----------------------------------------------------------
// | 以下函数为直接返回
// +-----------------------------------------------------------

/**
 * 返回对象
 * @param $array 响应数据
 * @example code:大于0为正确返回，反之错误返回。
 * 约定：
 * 1为默认返回，-1为默认错误
 * -1001~-1999 用户权限信息错误
 *   -1001 登录失效
 *   -1002 没有操作权限
 *   -1003 
 * -2001~-2999 系统错误
 *   -2001 发生了严重错误导致无法进行
 * -3001~-3999 数据错误
 *   -3001 数据结构错误
 *   -3002 数据存在冲突
 * -4001~-4999 配置错误
 */
function resultArray($array, $code = 1)
{
    if (isset($array['data'])) {
        $array['error'] = '';
    } elseif (isset($array['error'])) {
        if ($code == 1) $code = -1;
        $array['data'] = '';
    }
    return json([
        'code'  => $code,
        'data'  => $array['data'],
        'error' => $array['error']
    ]);
}

/**
 * 调试方法
 * @param  array   $data  [description]
 */
function p($data, $die = 1)
{
    echo "<pre>";
    print_r($data);
    echo "</pre>";
    if ($die) die;
}

/**
 * 用户密码加密方法
 * @param  string $str      加密的字符串
 * @param  [type] $auth_key 加密符
 * @return string           加密后长度为32的字符串
 */
function user_md5($str, $auth_key = '')
{
    return '' === $str ? '' : md5(sha1($str) . $auth_key);
}

/**
 * 获取系统配置
 * @param {string} configName: 配置项名，为空则返回所有配置
 * @return: {string}
 */
function getSystemConfig($configName = '')
{
    $data = cache('thinkvue');
    if (!$data) {
        $configs = model('SystemConfig')->select()->toArray();
        foreach ($configs as $k => $v) {
            $data[$v['name']] = $v;
        }
        cache('thinkvue', $data, 0);
    }
    if ($configName)
        if (isset($data[$configName]['value']))
            return $data[$configName]['value'];
        else
            throw new UnexpectedValueException('不存在该名称的配置');
    else
        return $data;
}

/**
 * 获取客户端类型
 * @return {int}
 */
function getClientType()
{
    static $clientTypes = ['OtherClient', 'miniprogram', 'MicroMessenger', 'QQ', 'AlipayClient', 'swan/', 'ToutiaoMicroApp'];
    // static $clientTypes =['other','微信小程序','微信','QQ小程序','支付宝小程序','百度小程序','头条小程序'];
    $client = $_SERVER['HTTP_USER_AGENT'];
    foreach ($clientTypes as $k => $v) {
        if (strpos($client, $v))
            return $k;
    }
    return 0;
}

/**
 * 获取客户端中文名称
 * @param {int} $clientType
 * @return {string}
 */
function getClientName(int $clientType)
{
    // static $clientTypes =['OtherClient','miniprogram','MicroMessenger','QQ','AlipayClient','swan/','ToutiaoMicroApp'];
    static $clientTypes = ['其他浏览器', '微信小程序', '微信', 'QQ小程序', '支付宝小程序', '百度小程序', '头条小程序'];
    return $clientTypes[$clientType];
}

// +-----------------------------------------------------------
// | 以下函数为标准返回['code'=>code,'msg'=>msg,'data'=>data]
// +-----------------------------------------------------------

/**
 * 获取授权token
 * @param string $code 通过get_authorize_url获取到的code
 * @param string $anonymous_code 限字节跳动小程序通过login获取到的anonymous_code
 * @return ['code'=>1,'token'=>token,'openid'=>openid,'msg'=>errMsg];
 */
function getTokenOpenid($code,$anonymous_code="")
{
    // $clientType=2; //TODO:调试
    $clientType=getClientType();
    if($clientType==1){ //微信小程序
        $appid = getSystemConfig('APPID_WXXCX');
        $appsecret = getSystemConfig('SECRET_WXXCX');
        $token_url = "https://api.weixin.qq.com/sns/jscode2session?appid={$appid}&secret={$appsecret}&js_code={$code}&grant_type=authorization_code";
        $data = http_send($token_url);
        if ($data['code'] == 200) {
            if($data['data']['errcode']!=0)
                return ['code'=>-4002, 'msg'=>$data['data']['errmsg']];
            return ['code'=>1,'session_key'=>$data['data']['session_key'],'openid'=>$data['data']['openid']];
        }
    }elseif($clientType==2){ //微信公众号
        $appid = getSystemConfig('APPID_WEIXIN');
        $appsecret = getSystemConfig('SECRET_WEIXIN');
        $token_url = "https://api.weixin.qq.com/sns/oauth2/access_token?appid={$appid}&secret={$appsecret}&code={$code}&grant_type=authorization_code";
        $data = http_send($token_url);
        if ($data['code'] == 200) {
            if(isset($data['data']['errcode']))
                return ['code'=>-4002, 'msg'=>$data['data']['errmsg']];
            return ['code'=>1,'token'=>$data['data']['access_token'],'openid'=>$data['data']['openid']];
        }
    }elseif($clientType==3){ //QQ
        $appid = getSystemConfig('APPID_QQ');
        $appsecret = getSystemConfig('SECRET_QQ');
        $token_url = "https://api.q.qq.com/sns/jscode2session?appid={$appid}&secret={$appsecret}&js_code={$code}&grant_type=authorization_code";
        $data = http_send($token_url);
        if ($data['code'] == 200) {
            if($data['data']['errcode']!=0)
                return ['code'=>-4002, 'msg'=>$data['data']['errmsg']];
            return ['code'=>1,'session_key'=>$data['data']['session_key'],'openid'=>$data['data']['openid']];
        }
    }elseif($clientType==4){ //支付宝 未支持
        $appid = getSystemConfig('APPID_ALIPAY');
        $appsecret = getSystemConfig('SECRET_ALIPAY');
    }elseif($clientType==5){ //百度
        $appid = getSystemConfig('APPID_BAIDU');
        $appsecret = getSystemConfig('SECRET_BAIDU');
        $data=['code'=>$code,'client_id'=>$appsecret,'sk'=>$appsecret];
        $token_url = "https://spapi.baidu.com/oauth/jscode2sessionkey ";
        $data = http_send($token_url,'POST',$data);
        if ($data['code'] == 200) {
            if(isset($data['data']['error']))
                return ['code'=>-4002, 'msg'=>$data['data']['error_description']];
            return ['code'=>1,'session_key'=>$data['data']['session_key'],'openid'=>$data['data']['openid']];
        }
    }elseif($clientType==6){ //头条
        $appid = getSystemConfig('APPID_BYTE');
        $appsecret = getSystemConfig('SECRET_BYTE');
        $data=['code'=>$code,'appid'=>$appsecret,'secret'=>$appsecret,'anonymous_code'=>$anonymous_code];
        $token_url = "https://api.weixin.qq.com/sns/oauth2/access_token?appid={$appid}&secret={$appsecret}&code={$code}&grant_type=authorization_code";
        $data = http_send($token_url);
        if ($data['code'] == 200) {
            if($data['data']['errcode']!=0)
                return ['code'=>-4002, 'msg'=>$data['data']['errmsg']];
            return ['code'=>1,
                    'session_key'=>$data['data']['session_key'],
                    'openid'=>$data['data']['openid'],
                    'anonymous_openid'=>$data['data']['anonymous_openid']
                ];
        }
    }
    return ['code'=>-1, 'msg'=>'非法请求，服务器拒绝服务'];
}

/**
 * 请求用户信息
 * @param 通过getTokenOpenid请求返回的access_token和openid
 * @return: {"openid":'',"nickname":'',"sex":"1","province":'',"city":'',"country":'',"headimgurl":url,"privilege":[]}
 */
function getUserinfo($access_token,$openid)
{
    $url="https://api.weixin.qq.com/sns/userinfo?access_token={$access_token}&openid={$openid}&lang=zh_CN";
    $data=http_send($url);
    if($data['code']==200)
        return json_decode($data['data'],true);
    return false;
}

/**
 * 发送HTTP请求 
 * @param String url: url地址 
 * @param String method: 请求模式， GET|POST default 'GET'
 * @param String data: 请求数据 default null
 * @param String headers: 模拟请求头 default NULL 
 * @param Boolean debug: 是否调试模式 default false
 * @param String url: url地址 
 * @param String url: url地址 
 * @return [$http_code, $return_data]
 */
function http_send($url, $method = 'GET', $data = null, $headers = array(), $debug = false)
{
    $ci = curl_init();
    /* Curl settings */
    curl_setopt($ci, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
    curl_setopt($ci, CURLOPT_CONNECTTIMEOUT, 30);
    curl_setopt($ci, CURLOPT_SSL_VERIFYPEER, FALSE);
    curl_setopt($ci, CURLOPT_SSL_VERIFYHOST, FALSE);
    curl_setopt($ci, CURLOPT_TIMEOUT, 30);
    curl_setopt($ci, CURLOPT_RETURNTRANSFER, true);

    switch ($method) {
        case 'POST':
            curl_setopt($ci, CURLOPT_POST, true);
            if (!empty($data)) {
                curl_setopt($ci, CURLOPT_POSTFIELDS, $data);
                $this->postdata = $data;
            }
            break;
    }
    curl_setopt($ci, CURLOPT_URL, $url);
    curl_setopt($ci, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ci, CURLINFO_HEADER_OUT, true);

    $response = curl_exec($ci);
    $http_code = curl_getinfo($ci, CURLINFO_HTTP_CODE);
    if ($debug) {
        echo "=====post data======\r\n";
        var_dump($data);

        echo '=====info=====' . "\r\n";
        print_r(curl_getinfo($ci));

        echo '=====$response=====' . "\r\n";
        print_r($response);
    }
    curl_close($ci);
    return ['code' => $http_code, 'data' => json_decode($response,true)];
}

/**
 * 发送短信（短信宝）
 * @param {string} mobile 手机号 
 * @param {string} content 短信内容
 * @return: 
 */
function sendmsg_smsbao($mobile,$content)
{
    if (empty($content) || empty($mobile)) 
        return ['code' => -2002 , 'msg'=>'参数不正确，短信内容和手机号缺一不可。'];
    $statusStr = [
        "0"  => "短信发送成功",
        "-1" => "参数不全",
        "-2" => "服务器空间不支持,请确认支持curl或者fsocket，联系您的空间商解决或者更换空间！",
        "30" => "密码错误",
        "40" => "账号不存在",
        "41" => "余额不足",
        "42" => "帐户已过期",
        "43" => "IP地址限制",
        "50" => "内容含有敏感词"
    ];
    $username = getSystemConfig('SMSBAO_USERNAME'); //短信平台帐号
    $password = getSystemConfig('SMSBAO_PASSWORD'); //短信平台密码（已用MD5加密)
    // $content = str_replace('{code}',$code,getSystemConfig('VERIFY_SMS_CONTENT')); //要发送的短信内容
    $url = "http://api.smsbao.com/sms?u={$username}&p={$password}&m={$mobile}&c={$content}";
    $httpData = http_send($url);
    if($httpData['code'] != 200)
        return ['code'=>-4003, 'msg'=>'服务器网络错误，请求失败'];
    $code = $httpData['data']==0 ? 1 : -2002;
    return ['code'=>$code,'msg'=>$statusStr[$httpData['data']]];
}

/**
 * 发送阿里云短信
 * @param {Object} data['code'=>6666] 短信模板里的变量
 * @param {string} mobile 手机号
 * @param {string} templateCode 短信模板号
 * @return: 标准返回
 */
function sendmsg_ali($data, $mobile,$templateCode)
{
    $accessKeyId=getSystemConfig('Ali_AccessKeyId');
    $accessSecret=getSystemConfig('Ali_AccessSecret');
    AlibabaCloud::accessKeyClient($accessKeyId, $accessSecret)->regionId('cn-hangzhou')->asDefaultClient();
    try {
        $result = AlibabaCloud::rpc()
            ->product('Dysmsapi')
            ->version('2017-05-25')
            ->action('SendSms')
            ->method('POST')
            ->host('dysmsapi.aliyuncs.com')
            ->options([ 'query' => [
                        'RegionId' => "cn-hangzhou",
                        'PhoneNumbers' => $mobile,
                        'SignName' => getSystemConfig('Ali_SignName'),
                        'TemplateCode' => $templateCode,
                        'TemplateParam' => json_encode($data)]])
            ->request();
        $result=$result->toArray();
        if($result['Message']=='OK')
            return ['code'=>1, 'msg'=>'OK'];
        else
            return ['code'=>-1, 'msg'=>$result['Message']];
    } catch (ClientException $e) {
        return ['code'=>-1,'msg'=>$e->getErrorMessage()];
    } catch (ServerException $e) {
        return ['code'=>-1,'msg'=>$e->getErrorMessage()];
    }
}

/**
 * 发送验证码
 * @param {string} $code 验证码
 * @param {string} $mobile 手机号 
 * @return: 标准返回
 */
function sendSmsVerify($code,$mobile)
{
    $smsInterface = getSystemConfig('SMS_Interface_Type');
    if($smsInterface==1) // 阿里云接口
        return sendmsg_ali(['code'=>$code],$mobile,getSystemConfig('Ali_Verify_TemplateCode'));
    elseif($smsInterface==2){
        $content=str_replace("{code}",$code,getSystemConfig('SMSBAO_VERIFY_SMS_CONTENT'));
        return sendmsg_smsbao($mobile,$content);
    }
}